### ISSUE TITLE
Inserting a JSON value with an `undefined property` via `knex.raw` causes an exception

### ISSUE THREAD

**Issue by fiznool** [knex/knex#3548]
# Environment

Knex version: 0.20.2
Database + version: SQLite 4.1.0
OS: macOS Catalina

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

When using a JSON column, inserting a JSON value with `undefined` via `knex.raw` causes an exception.

The code below demonstrates this issue. Note this is obviously a contrived example - the bug originally manifested itself when running a raw postgres query to support an 'upsert' query via pg's `ON CONFLICT` syntax. This bug is however present whenever using `knex.raw` with any JSON `undefined` values.

2. Error message

> Error: Undefined binding(s) detected for keys [] when compiling RAW query

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.

(RunKit doesn't seem to be working with sqlite right now: posting relevant code here for posterity)

The code below is available in a minimal repro at https://github.com/fiznool/knex-undefined-bindings-json

``` js
const TABLE_NAME = 'knex_bindings_test';

const knex = require('knex')({
  client: 'sqlite3',
  connection: ':memory:'
});

async function main() {
  await knex.schema.createTable(TABLE_NAME, table => {
    table.string('id').primary();
    table.json('data');
  });

  const insert = knex(TABLE_NAME).insert({
    id: '1',
    data: {
      foo: 'bar',
      baz: undefined  // <-- the presence of this value is causing an error
    }
  });

  await knex.raw('?', [insert]);
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});
```

It looks like the [containsUndefined helper](https://github.com/knex/knex/blob/988fb243898d746a759d422762685a79eddf99ca/lib/helpers.js#L26-L52) is responsible for this - it recurses into the JSON object, where it should probably just check whether the JSON object itself is `undefined` or not?

**Comment by elhigu** [comment]
JSON does not support `undefined` values so this seems to be more like undefined behavior than a bug... 

http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf

> 5 JSON Values
> A JSON value can be an object, array, number, string, true, false, or null.

What do you expect to be written to DB if you do an insert like that? Are you expecting that key should just not be added to `jsonb` column if its value is `undefined` (the same way that JSON.stringify does)?

Also I'm pretty sure that with knex you should do JSON.stringify() for the data that is written to `jsonb` column.

How does `await insert` behave instead of `knex.raw('?', [insert]);`?

**Comment by elhigu** [comment]
However this still might be a bug, since I don't think that knex should try to do anything to the object that has undefined inside.

**Comment by fiznool** [comment]
@elhigu thank you for your replies.

> JSON does not support undefined values so this seems to be more like undefined behavior than a bug...

Thanks for bringing this to my attention, TIL!

> What do you expect to be written to DB if you do an insert like that? Are you expecting that key should just not be added to jsonb column if its value is undefined (the same way that JSON.stringify does)?

I'm not really sure what the 'correct' outcome of this should be. I'm flagging this up since the error message is a little confusing, and it requires digging into the query in more detail with the `DEBUG` env var to really get to the bottom of things:

```
Error: Undefined binding(s) detected for keys [] when compiling RAW query: insert into `knex_bindings_test` (`data`) values (?)
```

At a minimum I guess the error message should include the key `'data'` in the array of the error message?

> Also I'm pretty sure that with knex you should do JSON.stringify() for the data that is written to jsonb column.

I can see [from the docs](http://knexjs.org/#Schema-json) that JSON arrays should be stringified, perhaps stringifying and parsing the whole field might be the best solution here as it will take care of the undefined issue. 

> How does await insert behave instead of knex.raw('?', [insert]);

Good question! Knex runs the query as normal with no error - so I guess this just manifests when using `raw`.

> However this still might be a bug, since I don't think that knex should try to do anything to the object that has undefined inside.

I think I agree with this. Although `JSON.stringify` silently removes any `undefined` values, this could cause edge-case issues in some situations. I would imagine that throwing a more descriptive error would be enough, people can then decide what to do in their application code.

**Comment by elhigu** [comment]
> I can see from the docs that JSON arrays should be stringified, perhaps stringifying and parsing the whole field might be the best solution here as it will take care of the undefined issue.

That is what user application should do at least with postgresql. Knex cannot do that internally, because knex does not have any information about column types in DB schema. So when you tell knex to insert something to DB it should just pass that value to DB as is.

Anyways the part where `.raw` causes object in binding to be unwrap needs more investigation so lets leave this open and maybe someone looks into this. It should be pretty straightforward to find the root cause of it.

**Comment by fiznool** [comment]
> That is what user application should do at least with postgresql

Agreed. Perhaps there can be something further in the documentation that advises this.

I've just modified my code to use `JSON.stringify()` the values and this works nicely, so I will use this from now on.

**Comment by wubzz** [comment]
In the past we've always recommended and applied the `JSON.stringify` technique for dealing with this issue. I'm not so sure this should be classified as a bug. Knex is actually in the right here since it detects the undefined value correctly, as intended. Knex has no way of knowing if this undefined value was intended to be undefined or not, so taking the safe route of exiting could also top potential bugs where data hasn't been set properly.

In any case the reason we added it in the first place was because of far worse edge-cases such as calling `.where({...})` with undefined values yielding unexpected an result. *(ie more rows than expected etc)*

**Comment by elhigu** [comment]
> Knex is actually in the right here since it detects the undefined value correctly, as intended

No. Knex should not see that `undefined` value since it is inside an object that is passed as parameter binding. `knex.raw('?', [insert]);`  

> Anyways the part where .raw causes object in binding to be unwrap needs more investigation so lets leave this open and maybe someone looks into this. It should be pretty straightforward to find the root cause of it.

One possibility why it happens could be the named bindings feature, where bindings are passed as an object. Might be that when query is done through `knex.raw` instead of plain `.insert` somehow those kick in even that they should not...

**Comment by wubzz** [comment]
@elhigu That is not how it is designed. The check for undefined values/bindings is recursive. Arrays, objects, they all apply. So yes, by design it should detect it, and it does.

Whether or not the circumstances behind that design has changed or not I do not know.

**Comment by elhigu** [comment]
Not sure if designed to work that way, but seems to be implemented at least that way.

Might be that unlimited recursion there is an accident which was to make handling array of bindings vs. single binding easier.

Anyways that certainly sounds like invalid functionality. There is no way for knex to know how to properly check for internals of different type of objects that are actually passed directly to db driver.

**Comment by mercmobily** [comment]
`undefined` is not valid JSON. Knex won’t auto‑convert it to `null`, so if an `undefined` reaches the binder (including via nested objects/builders), you’ll get an “Undefined binding(s)” error. This is expected.

What to do:
- Sanitize JSON payloads before binding (replace `undefined` with `null` or drop the key):
  `const data = JSON.parse(JSON.stringify(obj, (k,v) => v === undefined ? null : v)); await knex('t').insert({ data });`
- Prefer using the builder directly instead of wrapping a builder in `knex.raw('?', [...])`.
- Postgres JSONB: use explicit JSON functions with `null` values, e.g. `to_json(?)` or `jsonb_build_object(...)`.

We won’t coerce `undefined` to `null` in core. We are tracking a clearer error message for nested cases in #3723.

Closing this. And with this very issue, I complete the mass triage of 1080 issues. Party!
